jax/_src/scipy/special.py

```diff
@@ -724,82 +725,249 @@ def i1(x: ArrayLike) -> Array:
+ 4.84409658339962045305E7,
+ 1.11855537045356834862E10,
+ 2.11277520115489217587E12,
+ 3.10518229857422583814E14,
+ 3.18121955943204943306E16,
+ 1.71086294081043136091E18,
+]
 
-  Raises:
-    TypeError if `v` is not integer.
-    ValueError if elements of array `z` are not float.
-  """
-  z = jnp.asarray(z)
-  z, = promote_dtypes_inexact(z)
-  z_dtype = lax.dtype(z)
-  if dtypes.issubdtype(z_dtype, complex):
-    raise ValueError("complex input not supported.")
-
-  v = core.concrete_or_error(operator.index, v, 'Argument v of bessel_jn.')
-  n_iter = core.concrete_or_error(int, n_iter, 'Argument n_iter of bessel_jn.')
-
-  bessel_jn_fun = partial(_bessel_jn, v=v, n_iter=n_iter)
-  for _ in range(z.ndim):
-    bessel_jn_fun = vmap(bessel_jn_fun)
-  return jnp.moveaxis(bessel_jn_fun(z), -1, 0)
+def j1_small(x: ArrayLike) -> Array:
+    '''
+    Implementation of J1 for x < 5 
+    '''
+    z = x * x
+    w = jnp.polyval(jnp.array(RP1), z) / jnp.polyval(jnp.array(RQ1), z)
+    w = w * x * (z - jnp.array(Z1)) * (z - jnp.array(Z2))
+    return w
+
+def j1_large(x: ArrayLike) -> Array:  
+    '''
+    Implementation of J1 for x > 5 
+    '''  
+    w = 5.0 / x
+    z = w * w
+    p = jnp.polyval(jnp.array(PP1), z) / jnp.polyval(jnp.array(PQ1), z)
+    q = jnp.polyval(jnp.array(QP1), z) / jnp.polyval(jnp.array(QQ1), z)
+    xn = x - THPIO4
+    p = p * jnp.cos(xn) - w * q * jnp.sin(xn)
+    return p * SQ2OPI / jnp.sqrt(x)
+
+def bessel_j1(z: ArrayLike) -> Array:
+    """
+    Bessel function of the first kind of order one and a real argument 
+    - using the implementation from CEPHES, translated to Jax, to match scipy to machine precision.
+
+    Reference:
+    Cephes mathematical library.
+
+    Args:
+        x: The sampling point(s) at which the Bessel function of the first kind are
+        computed.
+
+    Returns:
+        An array of shape `x.shape` containing the values of the Bessel function
+    """
+
+    z = jnp.asarray(z)
+    z, = promote_dtypes_inexact(z)
+    z_dtype = lax.dtype(z)
+
+    if dtypes.issubdtype(z_dtype, complex):
+      raise ValueError("complex input not supported.")
+
+    return jnp.sign(z)*jnp.where(jnp.abs(z) < 5.0, j1_small(jnp.abs(z)),j1_large(jnp.abs(z)))
+
+def j0_small(x: ArrayLike) -> Array:
+    '''
+    Implementation of J0 for x < 5 
+    '''
+    z = x * x
+    # if x < 1.0e-5:
+    #     return 1.0 - z/4.0
+
+    p = (z - jnp.array(DR10)) * (z - jnp.array(DR20))
+    p = p * jnp.polyval(jnp.array(RP0),z)/jnp.polyval(jnp.array(RQ0), z)
+    return jnp.where(x<1e-5,1-z/4.0,p)
+    
+
+def j0_large(x: ArrayLike) -> Array:
+    '''
+    Implementation of J0 for x >= 5    
+    '''
+
+    w = 5.0/x
+    q = 25.0/(x*x)
+    p = jnp.polyval(jnp.array(PP0), q)/jnp.polyval(jnp.array(PQ0), q)
+    q = jnp.polyval(jnp.array(QP0), q)/jnp.polyval(jnp.array(QQ0), q)
+    xn = x - PIO4
+    p = p * jnp.cos(xn) - w * q * jnp.sin(xn)
+    return p * SQ2OPI / jnp.sqrt(x)
+
+def bessel_j0(z: ArrayLike) -> Array:
```

Comment:
Let's call this `j0`, as it's essentially a JAX implementation of  `scipy.special.j0`

Reply:
Sure thing.